### PR TITLE
Added contrib module for turn-based battle system

### PR DESCRIPTION
**Pull request by BattleJenkins**
This adds a new contrib file, a module that provides a simple framework for a tabletop style turn based battle system. From the module itself:

This is a framework for a simple turn-based combat system, similar
to those used in D&D-style tabletop role playing games. It allows
any character to start a fight in a room, at which point initiative
is rolled and a turn order is established. Each participant in combat
has a limited time to decide their action for that turn (30 seconds by
default), and combat progresses through the turn order, looping through
the participants until the fight ends.

Only simple rolls for attacking are implemented here, but this system
is easily extensible and can be used as the foundation for implementing
the rules from your turn-based tabletop game of choice or making your
own battle system.

#### Brief overview of PR changes/additions
Adds contrib module for a simple turn-based battle system.

#### Motivation for adding to Evennia
Combat systems are a near-universal feature for MUDs, and currently no contrib module provides a framework for them - this system provides a framework that resembles those of common tabletop role playing games and can be readily extended to suit the coder's needs.

#### Other info (issues closed, discussion etc)
I've tested this thoroughly on my local system and have fixed all the bugs I could find, but it hasn't been tested by multiple users yet.

### CODE REVIEW

File: evennia/contrib/turnbattle.py [hunk 1]
```diff
@@ -0,0 +1,520 @@
...
+        This performs the actual command.
+        """
+        here = self.caller.location
+        fighters = []
+        
+        if not self.caller.db.hp: # If you don't have any hp
+            self.caller.msg("You can't start a fight if you've been defeated!")
+            return
+        for thing in here.contents: # Test everything in the room to add it to the fight.
+            if thing.db.HP: # If the object has HP...
```
Comment (Griatch) [inline evennia/contrib/turnbattle.py]: This is a nice touch, it means you could fight a table or door assuming it has HP. :)

File: evennia/contrib/turnbattle.py [hunk 2]
```diff
@@ -0,0 +1,520 @@
...
+        """
+        for fighter in self.db.fighters:
+            combat_cleanup(fighter) #Clean up the combat attributes for every fighter.
+    
+    def at_repeat(self):
+        """
+        Called once every self.interval seconds.
+        """
+        currentchar = self.db.fighters[self.db.turn] # Note the current character in the turn order.
+        self.db.timer -= 1 # Count down the timer by one second.
```
Comment (Griatch) [inline evennia/contrib/turnbattle.py]: Ticking every second is a a simple but wasteful way to handle a countdown. The only times that the script cares about are after 20s (when the warning happens) and 30s (when the next round is force-started). You could therefore get away with ticking the script only every 10 seconds instead. You use the scripts' `restart` method whenever you want to prematurely start the next round on-demand. You can simply have your action-set commands check in to the script whenever an action is performed, so it knows when the actions are used up.

I'm not necessarily requiring you to rewrite this, but doing things on-demand is the recommended way to handle timeouts in an asynchronous system like Evennia. Checking things over and over for change is generally a bad practice.

File: evennia/contrib/turnbattle.py [hunk 3]
```diff
@@ -0,0 +1,520 @@
...
+        for fighter in self.db.fighters:
+            if fighter.db.Combat_LastAction != "disengage": # If a character has done anything but disengage
+                DisengageCheck = False
+        if DisengageCheck == True: # All characters have disengaged
+            self.obj.msg_contents("All fighters have disengaged! Combat is over!")
+            self.stop() # Stop this script and end combat.
+            return
+        
+        # Check to see if only one character is left standing. If so, end combat.
+        DefeatedCharacters = 0
```
Comment (Griatch) [inline evennia/contrib/turnbattle.py]: Python PEP8 standard (and thus Evennia's style guide) dictates that local variables are always written in lowercase with underscores. CamelCase is only ever used for class names.

File: evennia/contrib/turnbattle.py [hunk 4]
```diff
@@ -0,0 +1,520 @@
...
+    
+    
+
+"""
+----------------------------------------------------------------------------
+COMBAT FUNCTIONS START HERE
+----------------------------------------------------------------------------
+"""
+def roll_init(character):
+    """
```
Comment (Griatch) [inline evennia/contrib/turnbattle.py]: As elsewhere, google-style docstrings are needed.

**Comment by Griatch** [comment]
@BattleJenkins The concept of a unit test is simple - they are simply example calls of the individual components of a system. making sure a given function actually returns the result you expect to a given input. These are run automatiucally and makes sure your combat system keeps working even when other things inside Evennia change. If we were to change something in the API that the combat system relies on, the unit test will hopefully fail and tell us that we must update the combat system to match. The wiki page on unit tests should explain most of what you need to know, but don't be shy to ask if you have further questions!

File: evennia/contrib/turnbattle.py [hunk 5]
```diff
@@ -0,0 +1,663 @@
...
+    Returns:
+        initiative (int): The character's place in initiative - higher
+        numbers go first.
+        
+    Notes:
+        By default, does not reference the character and simply returns
+        a random integer from 1 to 1000.
+    """
+    return randint(1,1000)
+    """
```
Comment (Griatch) [inline evennia/contrib/turnbattle.py]: docstrings are usually at the top of the function/method/class etc. Probably better to put this info in the Notes section of the docstring instead or, if you want to have it in the  code, make it a normal comment with # markers.

File: evennia/contrib/turnbattle.py [hunk 6]
```diff
@@ -0,0 +1,663 @@
...
+        if self.db.timer == 10: # 10 seconds left
+            currentchar.msg("WARNING: About to time out!")
+        
+        # Force current character to disengage if timer runs out.
+        if self.db.timer <= 0:
+            self.obj.msg_contents("%s's turn timed out!" % currentchar)
+            spend_action(currentchar, 'all', action_name="disengage") # Spend all remaining actions.
+            
+            
+    def initialize_for_combat(self, character):
```
Comment (Griatch) [inline evennia/contrib/turnbattle.py]: Needs the `Args:` block.

**Comment by Griatch** [comment]
@BattleJenkins 

I ran the combat module with the latest fixes, seems to work fine. There are a few remaining issues:
1. When your turn times out, the other player only sees a message about the other timing out, there is no mention that it is now your turn. This is pretty clear on two players but could be a bit confusing for a larger fight. 
2. When trying to use an exit while in combat you get a traceback (this is a simple error, `self` on the object is the object itself, not the command):
       ```
Traceback (most recent call last):
 "/home/griatch/Devel/Home/evennia/evennia/evennia/objects/objects.py", line 683, in move_to
h     if not self.at_before_move(destination):
   File "/home/griatch/Devel/Home/evennia/evennia/evennia/contrib/turnbattle.py", line 308, in at_before_move
     self.caller.msg("You can't exit a room while in combat!")
 AttributeError: 'Character' object has no attribute 'caller'
        ```
3. It might be an idea to wrap your combat commands in a small cmdset. That way your install instructions could just be `from evennia.contrib import turnbattle` and then `self.add(turnbattle.BattleCmdSet)`.
4. I wonder if the combat should not offer some sort of custom help command. I reflexively entered help when going into combat and I figure many newbies would too - maybe help will just show a list of the available combat commands while help all and help <specific topic> will still show the normal help, something like that?

**Comment by Griatch** [comment]
I also see some `[EE] Script Combat Turn Handler(#1) of type 'TurnHandler': at_repeat() error ''NoneType' object has no attribute 'cancel''.` errors in the logs. I wonder if this is an issue in the TurnHandler jor in the scripting mechanism itself though.

**Comment by BlauFeuer** [comment]
The lgtm analysis doesn't seem to always run.  I don't think it's directly related to your code.  I've seen PRs accepted with the lgtm "could not build the merge commit" message, so wouldn't worry much about it.

**Comment by Griatch** [comment]
Yeah, as Blau said, the LGTM PR analysis seems to itself sometimes fail. Travis one is the important one for stability (once the unittests are in).

**Comment by Griatch** [comment]
@BattleJenkins Cool, at a quick glance it looks really nice now! I'll re-review it when the unit tests are in place. 

Btw, I turned off the LGTM Pull Request integration for now, it fails to load way too often on PRs for it to be useful.

**Comment by Griatch** [comment]
Looks good now, thanks for the contribution - merged with a rebase.

**Pull request closed without merging**